src/automatic_dantzig_wolfe.jl

```diff
@@ -0,0 +1,370 @@
+        return model.obj_dict[k]
+    elseif typeof(model.obj_dict[k]) <: Array
+        axs = axes(model.obj_dict[k])
+        return JuMP.Containers.DenseAxisArray(model.obj_dict[k], axs...)
+    else
+        # Can this case happen?
+        return error("Type of constraint reference can not be handled.")
+    end
+end
+
+# Add anonymous constraints and axes from the model to constraints_and_axes (car)
+function _add_anonymous_var_con!(car::Constraints_and_Axes, model::JuMP.Model)
+    types =  JuMP.list_of_constraint_types(model)
+    for t in types
+        if t[1] != VariableRef
+            for c in JuMP.all_constraints(model, t[1], t[2])
+                if !in(c, car.constraints)
+                    push!(car.constraints, c)
+                    car.constraints_to_axes[c] = Array{BlockDecomposition.Axis,1}()
+                    car.constraints_to_variables[c] = _get_variables_in_constraint(model, c)
+                end
+            end
+        end
+    end
+    for v in JuMP.all_variables(model)
+        push!(car.variables, JuMP.index(v))
+    end
+end
+
+function _add_constraint!(
+    o::Constraints_and_Axes,
+    c::JuMP.ConstraintRef,
+    model::JuMP.Model,
+    reference_constraints_name::T,
+) where T <: JuMP.Containers.DenseAxisArray
+    push!(o.constraints, c)
+    axes_of_constraint = _get_axes_of_constraint(reference_constraints_name)
+    for r in axes_of_constraint
+        push!(o.axes, r)
+    end
+    o.constraints_to_axes[c] = axes_of_constraint
+    o.constraints_to_variables[c] = _get_variables_in_constraint(model, c)
+end
+
+function _get_axes_of_constraint(
+    reference_constraints_name::T
+) where T <: JuMP.Containers.DenseAxisArray
+    axes_of_constraint = Array{BlockDecomposition.Axis,1}()
+    for a in reference_constraints_name.axes
+        if a != 1   # Axes of the form 1:1 do not matter (single constraints)
+            push!(axes_of_constraint, Axis(a))
+        end
+    end
+    return axes_of_constraint
+end
+
+function _get_variables_in_constraint(model::JuMP.Model, constraint::JuMP.ConstraintRef)
+    f = MOI.get(model, MathOptInterface.ConstraintFunction(), constraint)
+    variables = Set{MOI.VariableIndex}()
+    for term in f.terms
+        push!(variables, term.variable_index)
+    end
+    return variables
+end
+
+# Computes for the given axes two decomposition structures:
+# In the first one (block_structure0) all constraints *indexed by at least one* element from
+# axes are in the master, in the second one (block_structure1), all constraints *not indexed
+# by any* element from axes are in the master
+function get_block_structure(
+    axes::Array{<:Axis,1},
+    constraints_and_axes::Constraints_and_Axes,
+    model::JuMP.Model,
+)
+    vertices = [Set{JuMP.ConstraintRef}(), Set{JuMP.ConstraintRef}()]
+    master_constraints = [Set{JuMP.ConstraintRef}(),Set{JuMP.ConstraintRef}()]
+    for c in keys(constraints_and_axes.constraints_to_axes)
```

Comment:
I think you should find another name for `constraints_and_axes`. I feel like there are two distinct things inside this structure but it's more than that. This structure contains the information to do the automatic decomposition isn't it ?

Reply:
Yes, that's right. All the information it holds is extracted from the model and is later used in the automatic decomposition.
Most important for the automatic decomposition is the mapping from constraints to index sets/ axes and from constraints to variables.
Do you have any suggestion what might be a better name for this struct? Or should I maybe split it into two structs (one containing the constraints, variables, axes of the model, the other one containing the mappings)?

Reply:
I renames it to model_description. Do you think that's fine?